### PR TITLE
Use named error return for initProcess#start

### PR DESCRIPTION
**Pull request by tedyu**
In initProcess#start, there is deferred func :
```
	defer func() {
		if err != nil {
```
Since the error return is not named, the above check doesn't cover all the cases.
e.g. a few lines below:
```
	if _, err := io.Copy(p.messageSockPair.parent, p.bootstrapData); err != nil {
		return newSystemErrorWithCause(err, "copying bootstrap data to pipe")
```
The check in deferred func doesn't check either err or, the actual returned error.

This PR changes the error return to be named.

### CODE REVIEW

**Comment by tedyu** [comment]
@hqhq @cyphar
Mind taking a look ?

**Comment by tedyu** [comment]
The deferred func starting line 352 is duplicate with the one starting line 304.

I think the second deferred func can be removed.

**Comment by tedyu** [comment]
What happens is that the second deferred func is ineffective.
I checked every err variable after line 352: they are all local (newly declared) to the operation.

I will drop the second deferred func.

**Comment by kolyshkin** [comment]
@tedyu thanks for the patch! Coincidentally, I was looking into this part of code as well and ended up with a similar patch.

Some discussion about the code is provided at https://github.com/opencontainers/runc/pull/1916#pullrequestreview-169551062 (also, this is a PR to blame) and yet it doesn't explain the double defer.

I think the defer needs to be moved up a bit (before `p.intelRdtManager.Apply()` or even before `p.manager.Apply()`), otherwise, if `p.intelRdtManager.Apply()` fails, `p.manager.Destroy()` is not called.

@crosbymichael @giuseppe PTAL

**Comment by mrunalp** [comment]
LGTM

**Comment by mrunalp** [comment]
@AkihiroSuda ptal

**Comment by tedyu** [comment]
@kolyshkin 
I have lifted the deferred func earlier.
Please take a look.

**Comment by tedyu** [comment]
@kolyshkin
Please let me know if there is anything else to be addressed.

**Comment by thaJeztah** [comment]
@crosbymichael ptal 🤗

**Comment by crosbymichael** [comment]
LGTM

**Comment by tedyu** [comment]
@crosbymichael @thaJeztah 
Can this be merged ?

Thanks

**Comment by thaJeztah** [comment]
I'm not a maintainer in this repository; @AkihiroSuda @mrunalp PTAL?

File: libcontainer/process_linux.go [hunk 1]
```diff
@@ -290,6 +290,16 @@ func (p *initProcess) start() error {
 		p.process.ops = nil
 		return newSystemErrorWithCause(err, "starting init process command")
 	}
+	defer func() {
+		if retErr != nil {
+			// TODO: should not be the responsibility to call here
```
Comment (AkihiroSuda) [inline libcontainer/process_linux.go]: Is this TODO still correct?

Comment (tedyu) [reply libcontainer/process_linux.go]: @AkihiroSuda 
I have dropped this TODO.

**Comment by AkihiroSuda** [comment]
LGTM

**Comment by tedyu** [comment]
@AkihiroSuda 
Can you merge this ?

thanks

**Comment by tedyu** [comment]
@mrunalp 
Can you merge this ?

thanks

**Comment by tedyu** [comment]
Thanks @AkihiroSuda

**Comment by AkihiroSuda** [comment]
> Signed-off-by: zyu

Please use your full name next time

File: libcontainer/process_linux.go [hunk 2]
```diff
@@ -279,7 +279,7 @@ func (p *initProcess) waitForChildExit(childPid int) error {
 	return nil
 }
 
-func (p *initProcess) start() error {
+func (p *initProcess) start() (retErr error) {
```
Comment (cyphar) [inline libcontainer/process_linux.go]: Nit: Please call this `Err`.

**Pull request merged**